### ISSUE TITLE
Ability to send push notifications via browser in iOS / Android

### ISSUE THREAD

**Issue by loorm** [citizenos/citizenos-fe-old#287]
Considering, that a lot of our potential users are mobile based internet users, there have been questions from clients about our ability to send push notificiation to members of group or topic.

Also, it would be nice, if we could put a Citizen OS icon on users' mobile desktop.

Both can be achieved via web app and browser, making mobile app development unnecessary. However, we need to investigate, if and how mobile browsers can send push notifications. @tiblu thought, that he has seen this functionality being used by some websites.

**Comment by tiblu** [comment]
Very important to not disturb Users with "Do you want to receive notifications" every time User visits. It has to be opt-in and non-intrusive.

**Comment by ilmartyrk** [comment]
Some links for info and examples 
[Javascript implementation](https://blog.sessionstack.com/how-javascript-works-the-mechanics-of-web-push-notifications-290176c5c55d)
[Show Notifications doc](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification)
[IOS example](https://www.pubnub.com/blog/sending-ios-push-notifications-via-apns-javascript-using-apns-phonegap/)
[Android example](https://www.pubnub.com/blog/sending-android-push-notifications-via-gcm-javascript-using-phonegap/)

**Comment by tiblu** [comment]
Talked to @loorm, the main scope is native notifications on Android and iOS.
Also agreed to timebox this to 2h.

##  TLDR

* Web push notifications and iOS native app would cover most platforms.
* Web push notification UX is similar to native app.
* Web push notifications can deliver push notificactions to most platforms BUT NOT iOS Safari.
* Web push notifications are received ONLY when browser process is running, even if on other site or in the background. IF browser is not running, the delivery is delayed until browser is opened.
* To send push notifications to iOS with default Safari browser, we MUST have an iOS app created.
* Web push notifications are OK to use, if we implement the permissions request properly. Suggestions on that:
    * https://developers.google.com/web/fundamentals/push-notifications/permission-ux
    * https://developers.google.com/web/fundamentals/push-notifications/display-a-notification
* Time estimate to cover most platforms:
    * Web push notifications - 80h (that is backend time, does NOT include UX, localization etc resources)
    * iOS native app and it's notifications - NO ETA.

## Native notifications

### iOS

Sources:

* https://support.apple.com/en-us/HT201925 - Use notifications on your iPhone, iPad, and iPod touch
* https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html - Apple Push Notification service (APNs) overview

### Android

* Create Android app to show Citizen OS inside Web View - https://developer.android.com/reference/android/webkit/WebView

## Browser notifcations

Best article on the whole solution - https://developers.google.com/web/fundamentals/push-notifications.

TLDR:

* A browser subscribes to notifications via Push API.
* Notifications are ONLY delivered when the browser process is running, does not matter if the User is on your site or not. IF browser process is NOT running when the notification is triggered, it will be delivered once the browser is opened.
    *  > A service worker is a "special" JavaScript file. The browser can execute this JavaScript without your page being open. It can even execute this JavaScript when the browser is closed. A service worker also has API's, like push, that aren't available in the web page (i.e. API's that aren't available out of a service worker script). 
   * Source: https://developers.google.com/web/fundamentals/push-notifications/how-push-works#step_3_push_event_on_the_users_device
* Requesting notification permission (`Notification.requestPermission()`) does NOT allow custom message to be sent explaining WHY does the site request noitification permissions.
    * > One edge case that you need to handle is if the user clicks the 'Block' button. If this happens, your web app will not be able to ask the user for permission again. They'll have to manually "unblock" your app by changing its permission state, which is buried in a settings panel. Think carefully about how and when you ask the user for permission, because if they click block, it's not an easy way to reverse that decision.
* Push service - service that handles the registration and delivery of push notifications. This is selected by browser and NO NEED to implement by yourself. To emphasize - YOU CANNOT choose a push service, browser chooses it for you.
* NEED to implement API to store push service registration info. (`POST /api/save-subscription/`)
* Notification registration is per device NOT per User. Each user has to grant permissions to receive notifications on each of the devices used.

* NOT SUPPORTED on:
    * Push API - Safari on iOS, Internet Explorer (excl Edge), Webview for Android.
    * Noitification API - Safari on iOS, Internet Explorer (excl Edge), Webview for Android.

![image](https://user-images.githubusercontent.com/7405383/76754767-56a52500-678b-11ea-80c7-8995681096d0.png)

Sources:

* https://developer.mozilla.org/en-US/docs/Web/API/notification#Browser_compatibility
* https://developer.mozilla.org/en-US/docs/Web/API/notification
* https://developer.mozilla.org/en-US/docs/Web/API/Push_API
* https://developers.google.com/web/fundamentals/push-notifications/
* https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API
* https://developers.google.com/web/fundamentals/push-notifications/permission-ux
* https://developers.google.com/web/fundamentals/push-notifications/display-a-notification
* https://github.com/web-push-libs/web-push
* https://developers.google.com/web/fundamentals/primers/service-workers

Terms:

* Android web view - basically embedded browser inside an Android app - https://developer.android.com/reference/android/webkit/WebView
* Push service - service that handles the registration and delivery of push notifications. This is selected by browser and NO NEED to implement by yourself. To emphasize - YOU CANNOT choose a push service, browser chooses it for you.

**Comment by KatiVellak** [comment]
Legally reviewed, no impact.
